test/replication-luatest/gh_7590_replicaset_assert_test.lua

```diff
@@ -0,0 +1,93 @@
+local t = require('luatest')
+local server = require('luatest.server')
+local cluster = require('luatest.replica_set')
+
+local g = t.group('gh-7590')
+
+g.before_all(function(g)
+    g.cluster = cluster:new({})
+    g.cfg = {
+        replication_timeout = 0.1,
+        replication = {
+            server.build_listen_uri('r1'),
+            server.build_listen_uri('r3'),
+            server.build_listen_uri('r2'),
+        },
+    }
+
+    -- We need to specify uuid explicitly, so that rb tree traversal
+    -- wouldn't take errored applier (applier from r2 to r3) first.
+    g.cfg.instance_uuid = '00000000-0000-0000-0000-000000000001'
+    g.r1 = g.cluster:build_and_add_server({alias = 'r1', box_cfg = g.cfg})
+
+    g.cfg.instance_uuid = '00000000-0000-0000-0000-000000000002'
+    g.r2 = g.cluster:build_and_add_server({alias = 'r2', box_cfg = g.cfg})
+
+    g.cfg.instance_uuid = '00000000-0000-0000-0000-000000000003'
+    g.r3 = g.cluster:build_and_add_server({alias = 'r3', box_cfg = g.cfg})
+
+    g.cluster:start()
+    g.r2:exec(function()
+        box.schema.create_space('test'):create_index('pk')
+    end)
+
+    g.cluster:wait_for_fullmesh()
+    g.r3:wait_for_vclock_of(g.r2)
+    g.r1:wait_for_vclock_of(g.r2)
+end)
+
+g.after_all(function(g)
+    g.cluster:drop()
+end)
+
+g.test_replicaset_state_machine_assert_fail = function(g)
+    g.r1:exec(function(cfg)
+        -- Let's drop replication from r2 instance so that it cannot
```

Comment:
The comment is extremely misleading. Or rather the node naming. Somewhy your order is [r1, r3, r2] instead of [r1, r2, r3]. I spent quite some time trying to understand why `dropping r2` means `replication[3] = nil`. Can you change it?

Reply:
Replication array cannot be {rep_a, nil, rep_b}. It was done to workaround this one. Fixed, it's clearer now.

Reply:
> Replication array cannot be {rep_a, nil, rep_b}.

You could use table.remove for that. Anyway, your current version looks fine as well.